src/Components/Web.JS/src/Rendering/LogicalElements.ts

```diff
@@ -49,36 +51,71 @@ export function toLogicalRootCommentElement(start: Comment, end: Comment): Logic
   // |- *div
   // |- *component
   // |- *footer
-  if (!start.parentNode) {
-    throw new Error(`Comment not connected to the DOM ${start.textContent}`);
+  const { start, end } = descriptor;
+  const existingDescriptor = start[logicalRootDescriptorPropname];
+  if (existingDescriptor) {
+    if (existingDescriptor !== descriptor) {
+      throw new Error('The start component comment was already associated with another component descriptor.');
+    }
+    return start as unknown as LogicalElement;
   }
 
   const parent = start.parentNode;
+  if (!parent) {
+    throw new Error(`Comment not connected to the DOM ${start.textContent}`);
+  }
+
   const parentLogicalElement = toLogicalElement(parent, /* allow existing contents */ true);
   const children = getLogicalChildrenArray(parentLogicalElement);
-  Array.from(parent.childNodes).forEach(n => children.push(n as unknown as LogicalElement));
 
   start[logicalParentPropname] = parentLogicalElement;
-  // We might not have an end comment in the case of non-prerendered components.
+  start[logicalRootDescriptorPropname] = descriptor;
+  const startLogicalElement = toLogicalElement(start);
+
   if (end) {
-    start[logicalEndSiblingPropname] = end;
-    toLogicalElement(end);
+    // We need to make each element between the start and end comments a logical child
+    // of the start node.
+    const rootCommentChildren = getLogicalChildrenArray(startLogicalElement);
+    const startNextChildIndex = Array.prototype.indexOf.call(children, startLogicalElement) + 1;
+    let lastMovedChild: LogicalElement | null = null;
+
+    while (lastMovedChild !== end as unknown as LogicalElement) {
+      const childToMove = children.splice(startNextChildIndex, 1)[0];
+      if (!childToMove) {
+        throw new Error('Could not find the end component comment in the parent logical node list');
+      }
+      childToMove[logicalParentPropname] = start;
+      rootCommentChildren.push(childToMove);
+      lastMovedChild = childToMove;
+    }
```

Comment:
The updated DOM sync algorithm heavily relies on a sound logical element hierarchy, even for prerendered content, hence these changes. This wasn't really a requirement in the past because the only operation we did on prerendered content was remove it, and we had a special "logical end sibling" property to help with that case.

Reply:
The sheer amount of sophistication and complexity in this is a bit of an alarm. Perhaps I'm missing key scenarios but all I thought we needed to do was:

 * During DOM sync, when we're walking through a list of sibling elements, if we ever see an interactive component marker:
   * Treat that as a distinct new element type with its own separate logic for insert/update/delete (i.e., insertions/updates/deletions are operations on the corresponding interactive root component)
   * Then continue the walk by skipping to the node after the "end" marker, if there is one

I suggested that LogicalElement might be a helpful abstraction for seeing the DOM in a way that already has helpers for getting "next sibling" etc with respect to this virtual hierarchy. However we should only use that if it really is helpful and convenient.

The impression I get here is that interactive rendering's use of LogicalElement has pretty different requirements to the usage during DomSync. As such, trying to unify this into an all-encompassing system might be forcing a lot of complexity and perhaps would compromise on perf too (e.g., needing this new rule that "if a node is a LogicalElement, so are all its descendants").

I appreciate there isn't a lot of free time to finish this off! However we do have to keep maintaining `DomSync.ts` and `LogicalElement.ts` in the long run. TBH `LogicalElement` already was scary to me before this change because it's really intricate and has so many special cases that I can't remember, and I just trust that it's so battle-hardened by now that all these special cases do actually together in any combination that can actually occur. There would be a good case for not expanding it to accommodate DomSync if DomSync's requirements can actually be met with a simpler alternative.

What do you think?